packages/core/src/new-api/internal/execution/execution-engine.ts

```diff
@@ -0,0 +1,219 @@
+import { IgnitionError } from "../../../errors";
+import { isDeploymentExecutionState } from "../../../internal/utils/guards";
+import { isRuntimeValue } from "../../type-guards";
+import { DeploymentResult } from "../../types/deployer";
+import { ExecutionEngineState } from "../../types/execution-engine";
+import {
+  ExecutionStateMap,
+  ExecutionStatus,
+} from "../../types/execution-state";
+import {
+  ExecutionResult,
+  FutureRestart,
+  FutureStart,
+  JournalableMessage,
+} from "../../types/journal";
+import {
+  AccountRuntimeValue,
+  Future,
+  FutureType,
+  RuntimeValueType,
+} from "../../types/module";
+import { safeToString } from "../reconciliation/utils";
+
+import { executionStateReducer } from "./executionStateReducer";
+import { isExecutionResult, isOnChainAction, isOnchainResult } from "./guards";
+
+type ExecutionBatch = Future[];
+
+export class ExecutionEngine {
+  public async execute(state: ExecutionEngineState): Promise<DeploymentResult> {
+    const { batches, module } = state;
+
+    const futures = module.getFutures();
+
+    for (const batch of batches) {
+      // TODO: consider changing batcher to return futures rather than ids
+      const executionBatch = batch.map((futureId) =>
+        this._lookupFuture(futures, futureId)
+      );
+
+      const batchResult = await this._executeBatch(executionBatch, state);
+
+      if (batchResult.some((b) => b.type === "execution-failure")) {
+        return { status: "failed" };
+      }
+
+      if (batchResult.some((b) => b.type === "execution-hold")) {
+        return { status: "hold" };
+      }
```

Comment:
These returns loose data, you don't know what failed or hold later.

Reply:
Agreed. I have added todo items to deal with failed and hold in the restartability issue: #258